syntax = "proto3";

package cormoran.rip;

// Runtime Input Processor Messages
message InputProcessorInfo {
    string name             = 1;  // Short name (max 8 chars for BLE)
    uint32 scale_multiplier = 2;  // Scaling multiplier
    uint32 scale_divisor =
        3;  // Scaling divisor (multiply by scale_multiplier/scale_divisor)
    int32 rotation_degrees = 4;  // Rotation angle in degrees
    // Auto-mouse layer settings
    bool auto_mouse_enabled = 5;  // Whether auto-mouse is enabled
    uint32 auto_mouse_layer = 6;  // Target layer ID for auto-mouse
    uint32 auto_mouse_activation_delay_ms = 7;  // Delay before activating (ms)
    uint32 auto_mouse_deactivation_delay_ms = 8;  // Delay before deactivating (ms)
}

message ListInputProcessorsRequest {
    // Empty - requests list of all input processors
}

message ListInputProcessorsResponse {
    // Empty - use notification to report processors
}

message GetInputProcessorRequest {
    string name = 1;  // Name of the input processor
}

message GetInputProcessorResponse {
    InputProcessorInfo processor = 1;
}

message SetScaleMultiplierRequest {
    string name  = 1;  // Name of the input processor to update
    uint32 value = 2;  // New scaling multiplier
}

message SetScaleDivisorRequest {
    string name  = 1;  // Name of the input processor to update
    uint32 value = 2;  // New scaling divisor
}

message SetRotationRequest {
    string name = 1;  // Name of the input processor to update
    int32 value = 2;  // New rotation angle in degrees
}

message ResetInputProcessorRequest {
    string name = 1;  // Name of the input processor to reset
}

message SetAutoMouseRequest {
    string name = 1;  // Name of the input processor to update
    bool enabled = 2;  // Whether auto-mouse is enabled
    uint32 layer = 3;  // Target layer ID for auto-mouse
    uint32 activation_delay_ms = 4;  // Delay before activating (ms)
    uint32 deactivation_delay_ms = 5;  // Delay before deactivating (ms)
}

message SetScaleMultiplierResponse {
    // Empty - use notification to report changes
}

message SetScaleDivisorResponse {
    // Empty - use notification to report changes
}

message SetRotationResponse {
    // Empty - use notification to report changes
}

message ResetInputProcessorResponse {
    // Empty - use notification to report changes
}

message SetAutoMouseResponse {
    // Empty - use notification to report changes
}

message Request {
    oneof request_type {
        ListInputProcessorsRequest list_input_processors = 1;
        GetInputProcessorRequest get_input_processor     = 2;
        SetScaleMultiplierRequest set_scale_multiplier   = 3;
        SetScaleDivisorRequest set_scale_divisor         = 4;
        SetRotationRequest set_rotation                  = 5;
        ResetInputProcessorRequest reset_input_processor = 6;
        SetAutoMouseRequest set_auto_mouse               = 7;
    }
}

message ErrorResponse {
    string message = 1;
}

message Response {
    oneof response_type {
        ErrorResponse error                               = 1;
        ListInputProcessorsResponse list_input_processors = 2;
        GetInputProcessorResponse get_input_processor     = 3;
        SetScaleMultiplierResponse set_scale_multiplier   = 4;
        SetScaleDivisorResponse set_scale_divisor         = 5;
        SetRotationResponse set_rotation                  = 6;
        ResetInputProcessorResponse reset_input_processor = 7;
        SetAutoMouseResponse set_auto_mouse               = 8;
    }
}

// Notifications
message InputProcessorChangedNotification {
    InputProcessorInfo processor = 1;
}

message Notification {
    oneof notification_type {
        InputProcessorChangedNotification input_processor_changed = 1;
    }
}
